homeassistant/components/amcrest/config_flow.py

```diff
@@ -0,0 +1,233 @@
+    RESOLUTION_LIST,
+    STREAM_SOURCE_LIST,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class AmcrestFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a Amcrest config flow."""
+
+    VERSION = 2
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
+        """Get the options flow for this handler."""
+        return AmcrestOptionsFlowHandler(config_entry)
+
+    def __init__(self):
+        """Initialize the Amcrest config flow."""
+        self.device_config = {}
+        self.discovery_schema = {}
+        self.import_schema = {}
+
+    async def async_step_import(self, import_config):
+        """Import a config entry from configuration.yaml."""
+        _LOGGER.debug("start import AMCREST: %s", import_config)
+        return await self.async_step_user(import_config)
+
+    async def async_step_user(self, user_input=None):
+        """Handle a Amcrest config flow start.
+
+        Manage device specific parameters.
+        """
+        errors = {}
+
+        if user_input is not None:
+            try:
+                device = AmcrestChecker(
+                    self.hass,
+                    user_input[CONF_NAME],
+                    user_input[CONF_HOST],
+                    user_input[CONF_PORT],
+                    user_input[CONF_USERNAME],
+                    user_input[CONF_PASSWORD],
+                )
+
+                serial_number = device.serial_number.rstrip("\r\n")
+
+                await self.async_set_unique_id(serial_number)
+
+                self._abort_if_unique_id_configured(
+                    updates={
+                        CONF_HOST: user_input[CONF_HOST],
+                        CONF_PORT: user_input[CONF_PORT],
+                    }
+                )
+
+                self.device_config = {
+                    CONF_NAME: user_input[CONF_NAME],
+                    CONF_HOST: user_input[CONF_HOST],
+                    CONF_PORT: user_input[CONF_PORT],
+                    CONF_USERNAME: user_input[CONF_USERNAME],
+                    CONF_PASSWORD: user_input[CONF_PASSWORD],
+                }
+
+                return await self._create_entry()
+
+            except LoginError:
+                errors["base"] = "faulty_credentials"
+
+            except AmcrestError:
+                errors["base"] = "device_unavailable"
+
+        data = self.discovery_schema or {
+            vol.Required(CONF_NAME, default=DEFAULT_NAME): str,
+            vol.Required(CONF_HOST): str,
+            vol.Required(CONF_USERNAME): str,
+            vol.Required(CONF_PASSWORD): str,
+            vol.Required(CONF_PORT, default=DEFAULT_PORT): int,
+        }
+
+        return self.async_show_form(
+            step_id="user",
+            description_placeholders=self.device_config,
+            data_schema=vol.Schema(data),
+            errors=errors,
+        )
+
+    async def _create_entry(self):
```

Comment:
Why is this async?

Reply:
And why is it extracted as a method ?

Reply:
because i prefer use it in async way. Is an error?

I was inspired by the Axis integration that uses it in this way